File: frontend/aplikasi_tms/lib/services/analytics_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';

class DashboardStats {
  final int totalVehicles;
  final int activeVehicles;
  final int totalDrivers;
  final int activeDrivers;
  final int totalTrips;
  final int ongoingTrips;
  final int completedTrips;
  final double totalDistance;
  final int maintenanceDue;

  DashboardStats({
    required this.totalVehicles,
    required this.activeVehicles,
    required this.totalDrivers,
    required this.activeDrivers,
    required this.totalTrips,
    required this.ongoingTrips,
    required this.completedTrips,
    required this.totalDistance,
    required this.maintenanceDue,
  });

  factory DashboardStats.fromJson(Map<String, dynamic> json) {
    return DashboardStats(
      totalVehicles: json['total_vehicles'],
      activeVehicles: json['active_vehicles'],
      totalDrivers: json['total_drivers'],
      activeDrivers: json['active_drivers'],
      totalTrips: json['total_trips'],
      ongoingTrips: json['ongoing_trips'],
      completedTrips: json['completed_trips'],
      totalDistance: json['total_distance'].toDouble(),
      maintenanceDue: json['maintenance_due'],
    );
  }
}

class VehicleUtilization {
  final int vehicleId;
  final String registrationNumber;
  final int totalTrips;
  final double totalDistance;
  final double utilizationRate;

  VehicleUtilization({
    required this.vehicleId,
    required this.registrationNumber,
    required this.totalTrips,
    required this.totalDistance,
    required this.utilizationRate,
  });

  factory VehicleUtilization.fromJson(Map<String, dynamic> json) {
    return VehicleUtilization(
      vehicleId: json['vehicle_id'],
      registrationNumber: json['registration_number'],
      totalTrips: json['total_trips'],
      totalDistance: json['total_distance'].toDouble(),
      utilizationRate: json['utilization_rate'].toDouble(),
    );
  }
}

class AnalyticsService {
  static Future<DashboardStats> getDashboardStats() async {
    try {
      final response = await http.get(
        Uri.parse('${ApiConfig.baseUrl}/dashboard/stats'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return DashboardStats.fromJson(data['stats']);
      } else {
        throw Exception('Failed to load dashboard stats');
      }
    } catch (e) {
      throw Exception('Error: $e');
    }
  }

  static Future<List<VehicleUtilization>> getVehicleUtilization() async {
    try {
      final response = await http.get(
        Uri.parse('${ApiConfig.baseUrl}/dashboard/vehicle-utilization'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final List<dynamic> utilizationJson = data['utilization'] ?? [];
        return utilizationJson.map((json) => VehicleUtilization.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load vehicle utilization');
      }
    } catch (e) {
      throw Exception('Error: $e');
    }
  }
}

File: backend/internal/services/analytics_service.go
package services

import (
	"database/sql"
	"fmt"
	"log"

	"github.com/youruser/aplikasi-tms/backend/internal/middleware"
	"github.com/youruser/aplikasi-tms/backend/internal/models"
)

// Constants for better maintainability
const (
	DaysInMonth = 30.0
)

func GetDashboardStats(db *sql.DB) (*models.DashboardStats, error) {
	var stats models.DashboardStats

	// Single optimized query to get all stats at once
	query := `
		SELECT 
			(SELECT COUNT(*) FROM vehicles) as total_vehicles,
			(SELECT COUNT(*) FROM vehicles WHERE operational_status = 'active') as active_vehicles,
			(SELECT COUNT(*) FROM drivers) as total_drivers,
			(SELECT COUNT(*) FROM drivers WHERE status = 'available') as active_drivers,
			(SELECT COUNT(*) FROM trips) as total_trips,
			(SELECT COUNT(*) FROM trips WHERE status IN ('ongoing', 'in_progress')) as ongoing_trips,
			(SELECT COUNT(*) FROM trips WHERE status = 'completed') as completed_trips,
			(SELECT COALESCE(SUM(distance), 0) FROM trips WHERE distance IS NOT NULL) as total_distance,
			(SELECT COUNT(*) FROM vehicles WHERE next_maintenance_date <= CURRENT_DATE AND next_maintenance_date IS NOT NULL) as maintenance_due
	`

	err := db.QueryRow(query).Scan(
		&stats.TotalVehicles,
		&stats.ActiveVehicles,
		&stats.TotalDrivers,
		&stats.ActiveDrivers,
		&stats.TotalTrips,
		&stats.OngoingTrips,
		&stats.CompletedTrips,
		&stats.TotalDistance,
		&stats.MaintenanceDue,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to get dashboard stats: %v", err)
	}

	return &stats, nil
}

func GetVehicleUtilization(db *sql.DB) ([]models.VehicleUtilization, error) {
	query := `SELECT v.id, v.registration_number, 
			  COUNT(t.id) as total_trips,
			  COALESCE(SUM(t.distance), 0) as total_distance,
			  CASE 
			    WHEN COUNT(t.id) > 0 THEN (COUNT(t.id)::float / $1) * 100 
			    ELSE 0 
			  END as utilization_rate
			  FROM vehicles v
			  LEFT JOIN trips t ON v.id = t.vehicle_id
			  GROUP BY v.id, v.registration_number
			  ORDER BY total_trips DESC`

	rows, err := db.Query(query, DaysInMonth)
	if err != nil {
		return nil, fmt.Errorf("failed to get vehicle utilization: %v", err)
	}
	defer rows.Close()

	var utilizations []models.VehicleUtilization
	for rows.Next() {
		var u models.VehicleUtilization
		err := rows.Scan(&u.VehicleID, &u.RegistrationNumber, &u.TotalTrips, &u.TotalDistance, &u.UtilizationRate)
		if err != nil {
			log.Printf("Error scanning vehicle utilization: %v", middleware.SanitizeForLog(err.Error()))
			return nil, fmt.Errorf("failed to scan vehicle utilization: %v", err)
		}
		utilizations = append(utilizations, u)
	}

	// Check for errors during iteration
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("error during row iteration: %v", err)
	}

	return utilizations, nil
}